test/core/event_engine/thread_pool_test.cc

```diff
@@ -12,117 +12,204 @@
+  std::atomic<int> fork_count{0};
+  std::function<void()> inner_fn;
+  inner_fn = [&]() {
+    auto curr_runcount = runcount.load(std::memory_order_relaxed);
+    // exit when the right number of closures have run, with some flex for
+    // relaxed atomics.
+    if (curr_runcount >= expected_runcount) return;
+    if (fork_count.load(std::memory_order_relaxed) *
+            num_closures_between_forks <=
+        curr_runcount) {
+      // skip incrementing, and schedule again.
+      pool.Run(inner_fn);
+      return;
+    }
+    runcount.fetch_add(1, std::memory_order_relaxed);
+  };
+  for (int i = 0; i < expected_runcount; i++) {
+    pool.Run(inner_fn);
+  }
+  // simulate multiple forks at a fixed frequency
+  int curr_runcount = 0;
+  while (curr_runcount < expected_runcount) {
+    absl::SleepFor(fork_freqency);
+    curr_runcount = runcount.load(std::memory_order_relaxed);
+    int curr_forkcount = fork_count.load(std::memory_order_relaxed);
+    if (curr_forkcount * num_closures_between_forks > curr_runcount) {
+      continue;
+    }
+    pool.PrepareFork();
+    pool.PostforkChild();
+    fork_count.fetch_add(1);
+  }
+  ASSERT_GE(fork_count.load(), expected_runcount / num_closures_between_forks);
+  // owners are the local pool, and the copy inside `inner_fn`.
+  pool.Quiesce();
 }
 
-// This can be re-enabled if/when the thread pool is changed to quiesce
-// pre-fork. For now, it cannot get stuck because callback execution is
-// effectively paused until after the post-fork reboot.
-TEST(ThreadPoolDeathTest, DISABLED_CanDetectStucknessAtFork) {
-  ASSERT_DEATH_IF_SUPPORTED(
-      [] {
-        gpr_set_log_verbosity(GPR_LOG_SEVERITY_ERROR);
-        ThreadPool p;
-        ScheduleSelf(&p);
-        std::thread terminator([] {
-          std::this_thread::sleep_for(std::chrono::seconds(10));
-          abort();
-        });
-        p.PrepareFork();
-      }(),
-      "Waiting for thread pool to idle before forking");
+void ScheduleSelf(ThreadPool* p) {
+  p->Run([p] { ScheduleSelf(p); });
 }
 
-void ScheduleTwiceUntilZero(ThreadPool* p, int n) {
+void ScheduleTwiceUntilZero(ThreadPool* p, std::atomic<int>& runcount, int n) {
+  runcount.fetch_add(1);
   if (n == 0) return;
-  p->Run([p, n] {
-    ScheduleTwiceUntilZero(p, n - 1);
-    ScheduleTwiceUntilZero(p, n - 1);
+  p->Run([p, &runcount, n] {
+    ScheduleTwiceUntilZero(p, runcount, n - 1);
+    ScheduleTwiceUntilZero(p, runcount, n - 1);
   });
 }
 
-TEST(ThreadPoolTest, CanStartLotsOfClosures) {
-  ThreadPool p;
+TYPED_TEST(ThreadPoolTest, CanStartLotsOfClosures) {
+  TypeParam p(8);
+  std::atomic<int> runcount{0};
   // Our first thread pool implementation tried to create ~1M threads for this
   // test.
-  ScheduleTwiceUntilZero(&p, 20);
+  ScheduleTwiceUntilZero(&p, runcount, 20);
+  p.Quiesce();
+  ASSERT_EQ(runcount.load(), pow(2, 21) - 1);
+}
+
+TYPED_TEST(ThreadPoolTest, ScalesWhenBackloggedFromSingleThreadLocalQueue) {
+  int pool_thread_count = 8;
```

Comment:
Is there an upper limit on the number of threads spawned by the thread pool ? If not, can you create some k* pool_thread_count callbacks in line 161 and line 187 to verify the lifeguard is able to create more and more threads as necessary ?

Reply:
There is no upper limit. As is, this test ensures that scaling happens, which is essential. I'm not sure if we can safely specify a minimum scale amount to assert for all implementations. Do you think some N threads would make this a better test?

The biggest issue with "more and more (N) threads" is that the pool implementation creates at most 1 thread per second, so the test could take a while depending on N.